### ISSUE TITLE
Should we switch from PHP to Node.JS or GOlang?  PLEASE REPLY ASAP

### ISSUE THREAD

**Issue by spencersnygg** [spencersnygg/VoteWise#14]
Alec, one of the programmers really wants to switch to GOlang.  "GO", also know as "Golang" or less common "Google language", appears to be a bit of a mash-up of Pascal and C++.  It is faster and can run concurrently instead of single stream.  It has been around for like 4 years and is very popular in China, but I'm not sure about the USA.

At this moment, I have an open mind since we haven't written any back-end code yet, other than the databases.  My biggest concern is that It's been a lot easier to find PHP guys, so if we switch out of PHP, and we lose the one guy we have who knows GO, we could be a bit screwed.  My understanding about Node is that, like much of Javascript, there are a million ways to skin a cat, so just because two people know Node, it doesn't mean that it's easy to figure out each other's code - making it harder to pass the baton from one part time volunteer to another.  But I could be wrong on this.

Whatever we go with, the decision has to be made pronto because we are really close to our deadline of Oct 1st.

Alec is a proponent of GO and I will let him explain the differences as he sees them:

"I've used PHP, Node.js, and Go in production webserver, so I'll give my
analysis on the pros/cons of each that could spur discussion.

PHP Pros

   - Pretty well known.  This has a potential downside in that it's the
   kind of language where, even if a lot of people know it, a higher
   percentage of PHP programmers are less competent, since it's such a common
   entry level language.  Not something that concerns me though.
   - Development cycle is easy.  Change code, refresh.
   - With HHVM we could also integrate Hack code with the rest of the PHP
   seamlessly, which is a type variant of PHP made by Facebook.  I don't know
   how well supported this is yet though.
   - Used effectively in production by several companies; Facebook is the
   best example.  Although they seem to be trying to move away from PHP as
   much as they can (like HHVM and Hack; also a lot of their code is C++
   nowadays).

PHP Cons

   - Slow.  Even HHVM is slower than Node/Go.
   http://dan.hersam.com/2015/02/25/go-vs-node-vs-php-vs-hhvm-and-wordpress-benchmarks/
   compares Go, Node, PHP, and PHP via HHVM.  This is partly because resources
   like database connections have to be reloaded every time a page is
   requested.
   - Standard library is poorly designed/buggy.  This is a common topic for
   PHP developers.  See https://www.reddit.com/r/lolphp/top/?sort=top&t=all
   for examples.  At Balihoo there was generally a lot of complaints in this
   regard.
   - Implicit typecasting and default settings are known to introduce
   security bugs/weird issues.
   - Lack of static typing in general doesn't work well for big projects.
   - No parallelism except per-request.  Meaning if we're doing I/O in a
   single request/response (like accessing the database), we can only do one
   database request/filesystem access at a time even when we could be doing
   multiple at the same time.  Not a big deal though.

Plus I'm concerned about the troubles we had setting up homestead; it was
supposed to make development easier, not harder to set up.  Although this
is not necessarily an issue with Laravel or PHP itself.  If we continue
having trouble with Homestead I'd probably prefer to forego it and develop
PHP directly without a VM.

The VM also consumed a fair amount of resources on my computer (it was
eating up 30% CPU at one point, not sure why), so I'd *prefer* being able
to develop in my normal environment.

Node.js Pros

   - Javascript is very well known.  Better than PHP even.  Front-end
   developers would automatically know the language for the back end.
   - Used effectively in production.  Examples of companies that use it
   include Netflix, New York Time, Paypal, Medium, LinkedIn, Groupon.
      - https://www.quora.com/What-companies-are-using-Node-js-in-production
      - Same language on front end and back end.  Can make it easier to
   share code, although this isn't necessarily useful for our needs.
   - We already have Node/NPM in our source tree for Gulp and such.
   - Pretty quick.  Not as fast as Go, but a lot faster than PHP.  Can, to
   an extent, take advantage of parallelism with I/O.  Not as well as Go
   though, since Javascript only runs single threaded.
   - I/O parallelism.  Better than PHP's per-request parallelism, but
   unfortunately the entire server runs in a single thread, which can still be
   wasteful and introduce unnecessary latency considering how many cores most
   machines have.
   - Better module system than PHP, especially with NPM.

Node.js Cons

   - Callback hell.  http://callbackhell.com/ for an example.  A downside
   to Node.js's approach to concurrency.
   - Lack of static typing can make large projects less manageable over
   time.  Something I dealt with at Balihoo.

Go Pros

   - Fast.  Faster than the other two for sure.
   - True parallelism is easy and avoids callback hell, so we can take full
   advantage of our server's resources and keep latency manageable.
   - Used effectively in production.  DropBox, CloudFlare, SoundCloud, and,
   of course, Google.
      -
      https://www.quora.com/What-systems-in-production-are-written-in-Go-if-any
      - https://blog.golang.org/go-at-io-frequently-asked-questions
   - Statically typed.  Much easier to keep a big project manageable.
   - Well designed standard library.  From personal experience, from best
   to worst standard libraries, I'd say: Go, Node.js, PHP.
   - Syntax/code is easy to learn and understand.  More self-explanatory
   than Node or PHP in my opinion.

Go cons

   - Less well known than Node.js at least, I'm not sure what kind of
   userbase it has.  It's well known in the circles I frequent (Balihoo was
   considering it for production), but not as many actually knew it.  The
   community is certainly productive at least, I've never hard a problem
   finding a library for something.

I'd be happy to answer any questions you have about comparing the three
languages and their environments.  I want to choose a language that suits
the whole team and project best, and hopefully this information helps us
discuss it.  I'd prefer Go over Node.js or PHP, but if it's not well known
enough, then Node.js sounds like a solid choice.

Thanks,

Alec"

**Comment by rovaughn** [comment]
> My understanding about Node is that, like much of Javascript, there are a million ways to skin a cat, so just because two people know Node, it doesn't mean that it's easy to figure out each other's code - making it harder to pass the baton from one part time volunteer to another. But I could be wrong on this.

I suppose that's a fair analysis of it, there are a TON of Node.js libraries and more than a few ways of doing things.  Although I wouldn't worry too much about getting everyone to understand the same code base.

> Alec is a proponent of GO and I will let him explain the differences as he sees them

Just to be clear, although I do have a preference for Go, it's not too strong, especially if I'd be the only one that knows it well.  I'd be more than happy to use Node.js or even PHP if that's really best for the group.

**Comment by spencersnygg** [comment]
Here is the conversation:

I asked Alec if one was easier to learn than the other.

ALEC: I would say which one's easier to learn depends on the languages you
already know.  However, I'm inclined to say Go is if you don't know
either.  The standard library just makes more sense and is less weird.  The
syntax is also super simple.




Matias responded to my question about his preferences of PHP vs Node vs GO

MATIAS: GO is not well known. I've only read about it but have not used it at all.
I think it's from Google. Node would be a good choice.  I would suggest
using the MEAN stack.  Especially since we already have AngularJS. It would
be a full stack with JS as the language for the entire development build



He later chimed in:

MATIAS: BTW I have Laravel running on my laptop, but version 4.  I chose that
version because of Homestead ' s issues.  We can try that version if
everyone is having a hard time with the installation.   Though, I would
prefer going with MEAN stack. Ether way would work.



I sent Matias's first comment to Alec who responded:

ALEC:  Alright, given everyone's backgrounds, it sounds like Node is the best
choice.  I agree with Matias, I think it'll fit in well with the rest of
the stack.
M.E.A.N. (MongoDB/Express.js/Angular.js/Node.js) makes sense, although we
should discuss the M part.  For our needs, a relational database like MySQL
or PostgreSQL makes more sense to me.


James chimed in:
JAMES:BTW, I was working with Kyle today, and told him about the possible
language switch on the backend. He said he would be eager to get his hands
dirty with Go, and he also knows Node.

**Comment by realisation** [comment]
Cool. I'd prefer to go with NODE, too. I've had a few discussions with seasoned people, and they say the languages that are here to stay are C++, Java, Javascript and Python, which obviously phases out PHP. So why develop something that's on its way to being archaic? I think I'd have more curiosity in doing the MEAN stack.

I'm wondering: is there a way to program concurrency into JS? Like, I know there are some Python libraries that allow you to use C++ for particular tasks - does that apply at all to a Node server situation?

**Comment by varNinja** [comment]
I would have chimed in earlier, but I was driving home :)

- The MEAN stack is common but just as common is the availability of Node.js <--> Relational Database plugins. (Example: https://www.npmjs.com/package/mysql) 

- I also admit that, when it comes to Javascript and Go, Javascript is definitely a lot more weird.

- Alec, does Go have a sort of library/package manager similar to NPM?

- Like James mentioned, I am very familiar with Javascript (as well as many of the other developers here it seems, which would make for some quicker development time), but I would be willing to do Go if we decided that was best for the project.

- Final note: Of course there are many ways to implement a Node.js back-end, but we can collectively decide the right way to code the back-end. If we decide certain conventions, stick to them, and fully document them in the repo, there should be no issues with passing the baton on.

**Comment by spencersnygg** [comment]
The discussion seems to have stopped before all ideas were explored.  Please continue.

**Comment by varNinja** [comment]
For the sake of coming to a final decision, I'm going to say that, in my opinion, Node.js is our better option. We will find tons of people for the project using Node.js as compared to Go, and that may be better for the health of the project going forward. I would absolutely love to get my hands dirty in Go, and I think it should be an option in the future, but for now, we have a deadline to meet and our strength is Node.js.

**Comment by spencersnygg** [comment]
Does NOde have the ability to make notation within the code like HTML does?

**Comment by varNinja** [comment]
If you mean like comments then yes, because all Node.js documents are just Javascript documents.

**Comment by spencersnygg** [comment]
Ok, Node it is.

**Comment by varNinja** [comment]
Any final comments from @rovaughn @realisation ?

**Comment by spencersnygg** [comment]
Someone just suggested http://expressjs.com/ as a Node framework with an easy learning curve if we need one.  Now for the Mongo vs MySQL......  Thoughts?

**Comment by varNinja** [comment]
Yeah I actually thought Express was a given (otherwise, we would be creating an HTTP server from scratch), so we should definitely use Express. 
(I'll edit this once I figure out how to answer the Mongo vs. MySQL argument)

**Comment by rovaughn** [comment]
I agree with @varNinja's points regarding Node.js as better for meeting our deadline and making sure new developers can be found if needed.  I think we feel about the same regarding Node vs Go.  And like he said, there's not a particular reason to prefer over MongoDB over any other database, since the interop with the relational DBs are just as easy.

Also, to answer @varNinja's question wrt NPM and Go, it's built into the go tool.  For instance if in a file you have

```go
import "github.com/lib/pq"
```

Then you just run `go get` in the directory and, if that package isn't already downloaded, it downloads it and installs it.  The package becomes available as `pq` in the rest of the file.

And @spencersnygg, just so you are aware, all of the languages we are looking at have the ability to have comments like HTML does.  In fact, I don't know of a language without comments. :)

Anyway, in conclusion, I'm for Node.js as our backend language.

**Comment by msalas74** [comment]
I believe that Node was built with C++, so I assume it would be.

**Comment by spencersnygg** [comment]
About Mongo vs MySQL, Matias said:

"MATIAS:MongoDB is much easier to program for, especially for those that already know JS. 
MySQL requires for you to have your schema established compared to MongoDB's ability
to dynamically change in structure.  That makes it more scalable and adaptable."

HE LATER ADDED:

"As a side note - my original code for the mockup already had some of the MEAN stack
development dependancies, like Node, and AngularJS with Gulp for dev task
management.  There's tons of online frameworks for MEAN for you guys to start with."


Matias' schedule just got really busy again, so he won't be able to help much.

**Comment by rovaughn** [comment]
We decided on Node.
